cantools/database/utils.py

```diff
@@ -135,41 +130,39 @@ def decode_data(data: bytes,
     }
 
     if allow_truncated and actual_length < expected_length:
-        # remove fields that are outside available data bytes
-        valid_bit_count = actual_length * 8
-        for field in fields:
-            if field.byte_order == "little_endian":
-                sequential_startbit = field.start
+        # remove signals that are outside available data bytes
+        actual_bit_count = actual_length * 8
+        for signal in signals:
+            if signal.byte_order == "little_endian":
+                sequential_start_bit = signal.start
             else:
-                # Calculate startbit with inverted indices.
+                # Calculate start bit with inverted indices.
                 # Function body of ``sawtooth_to_network_bitnum()``
                 # is inlined for improved performance.
-                sequential_startbit = (8 * (field.start // 8)) + (7 - (field.start % 8))
+                sequential_start_bit = (8 * (signal.start // 8)) + (7 - (signal.start % 8))
 
-            if sequential_startbit + field.length > valid_bit_count:
-                del unpacked[field.name]
+            if sequential_start_bit + signal.length > actual_bit_count:
+                del unpacked[signal.name]
 
-    decoded = {}
-    for field in fields:
+    decoded: Dict[str, SignalValueType] = {}
+    for signal in signals:
         try:
-            value = unpacked[field.name]
-
-            if decode_choices:
-                try:
-                    decoded[field.name] = field.choices[value]  # type: ignore[index]
-                    continue
-                except (KeyError, TypeError):
-                    pass
-
-            if scaling:
-                decoded[field.name] = field.scale * value + field.offset
-                continue
-            else:
-                decoded[field.name] = value
-
+            value = unpacked[signal.name]
         except KeyError:
             if not allow_truncated:
                 raise
+            continue
+
+        if decode_choices and signal.choices is not None and value in signal.choices:
```

Comment:
Maybe we could simplify this too like in the encoding function, if the performance allows it:
```python3
if scaling:
    decoded[signal.name] = signal.raw_to_scaled(value, decode_choices)
else:
    decoded[signal.name] = value
```

Reply:
yes, if we decide that choices are only considered if `scaling` is `True` and get rid of the `decode_choices` parameter. (I'm a bit on the line about whether this is a good idea.)